Authentication session with changelog transaction

graziang opened the pull request:
Closes #23881
Closes #32658

Authentication sessions now use the changelog transactions implemented in both `InfinispanAuthenticationSessionProvider` and `RemoteInfinispanAuthenticationSessionProvider`. This way, concurrency write issues for auth sessions should be resolved. Thanks @pruivo  for the help.

graziang committed:
Auth session with changelog transaction

Closes #23881
Closes #32658

Signed-off-by: Giuseppe Graziano <g.graziano94@gmail.com>

pruivo commented on testsuite/model/src/main/java/org/keycloak/testsuite/model/HotRodServerRule.java:
```diff
@@ -104,6 +105,9 @@ private void createKeycloakCaches(boolean async, String... cache) {
                 .site("site-1").backupFailurePolicy(BackupFailurePolicy.IGNORE).strategy(BackupConfiguration.BackupStrategy.SYNC)
```
@graziang Set the backup failure policy `BackupFailurePolicy.FAIL` here and in line 102. It ignores the deadlock and assumes the update is applied.

pruivo commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/AuthenticationSessionAdapter.java:
```diff
@@ -25,6 +25,7 @@
 import org.keycloak.models.UserModel;
 import org.keycloak.models.light.LightweightUserAdapter;
 import org.keycloak.models.sessions.infinispan.entities.AuthenticationSessionEntity;
+import org.keycloak.models.sessions.infinispan.entities.RootAuthenticationSessionEntity;
```
unused import

pruivo commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/AuthenticationSessionAdapter.java:
```diff
@@ -45,19 +46,22 @@
 
     private final KeycloakSession session;
-    private final RootAuthenticationSessionAdapter parent;
+    private final  SessionEntityUpdater<AuthenticationSessionEntity> updater;;
```
nitpicking, to keep the diff easy to read, I would keep the `parent` field here and add the `updater` below.

pruivo commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/AuthenticationSessionAdapter.java:
```diff
@@ -45,19 +46,22 @@
-    private final RootAuthenticationSessionAdapter parent;
+    private final  SessionEntityUpdater<AuthenticationSessionEntity> updater;;
     private final String tabId;
     private final AuthenticationSessionEntity entity;
```
can be removed. You can use `updater.getEntity()`.

pruivo commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanAuthenticationSessionProvider.java:
```diff
@@ -23,47 +23,66 @@
 
-        this.tx = new InfinispanKeycloakTransaction();
+        this.cache = cache;
+        this.sessionTx = new InfinispanChangelogBasedTransaction<>(session, cache, remoteCacheInvoker, SessionTimeouts::getAuthSessionLifespanMS, SessionTimeouts::getAuthSessionLifespanMS, serializer);
```
Set max-idle to -1 if you don't need max-idle expiration. Max-idle has a performance impact since each read needs to update the access time everywhere.

pruivo commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanAuthenticationSessionProvider.java:
```diff
@@ -78,21 +97,23 @@ public RootAuthenticationSessionModel createRootAuthenticationSession(RealmModel
-        tx.put(cache, id, entity, expirationSeconds, TimeUnit.SECONDS);
-
+        //tx.put(cache, id, entity, expirationSeconds, TimeUnit.SECONDS);
+        SessionUpdateTask<RootAuthenticationSessionEntity> createLoginFailureTask = Tasks.addIfAbsentSync();
```
nitpicking, typo: `createLoginFailureTask`?  :) You are creating a session.

pruivo reviewed (changes requested):
I have some suggestions :)

graziang commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanAuthenticationSessionProviderFactory.java:
```diff
@@ -163,24 +220,8 @@ private static void updateAuthSession(RootAuthenticationSessionEntity rootAuthSe
-                }
-            }
-        }
+        this.authSessionsCache.replace(rootAuthSession.getId(), new SessionEntityWrapper<>(rootAuthSessionWrapper.getLocalMetadata(), rootAuthSession));
```
@pruivo Can this work?

pruivo replied:
Yes and no. I don't fully understand this method but it is invoked from an event notification. All KC nodes in the cluster will invoke replace at the same time... it works but it is not good for performance reasons :)

graziang commented on testsuite/model/src/test/java/org/keycloak/testsuite/model/session/AuthenticationSessionTest.java:
```diff
@@ -198,4 +200,58 @@ public void testRemoveExpiredAuthSessions() {
+
+        withRealm(realmId, (session, realm) -> {
+            RootAuthenticationSessionModel rootAuthSession = session.authenticationSessions().getRootAuthenticationSession(realm, rootId);
+            //TODO: rootAuthSession should be null not just empty check removeAuthenticationSessionByTabId
```
@pruivo We need to figure out how to remove the root auth session when there are no more auth sessions. The check should happen after the update tasks are finished. See `removeAuthenticationSessionByTabId` method

graziang commented:
@pruivo Thanks for the review! I applied your suggestions and refactored a bit. I left few comments on some doubts I have.

pruivo commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/remote/updater/authsession/RootAuthenticationSessionUpdater.java:
```diff
@@ -0,0 +1,241 @@
+            return null;
+        }
+        changes.forEach(c -> c.accept(cachedEntity));
+        return cachedEntity;
```
@graziang For multi-site deployment, this should do the trick to remove the entry from the remote cache when the map is empty. 
```suggestion
        return cachedEntity.getAuthenticationSessions().isEmpty() ? null : cachedEntity;
```

graziang replied:
Yeah it works, We would need something similar for `RootAuthenticationSessionAdapter`.

pruivo commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanAuthenticationSessionProviderFactory.java:
```diff
@@ -121,32 +176,34 @@ protected void eventReceived(InfinispanAuthenticationSessionProvider provider, R
         if (! (clEvent instanceof AuthenticationSessionAuthNoteUpdateEvent event)) {
             return;
         }
 
```
@graziang to limit the event handling to the primary owner of the entry
```java
        var distribution = authSessionsCache.getAdvancedCache().getDistributionManager();
        if (distribution != null && !distribution.getCacheTopology().getDistribution(event.getAuthSessionId()).isPrimary()) {
            // Distribution is null for non-clustered caches (local-cache, used by start-dev mode).
            // If not the primary owner of the key, skip event handling.
            return;
        }
```

pruivo commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanAuthenticationSessionProviderFactory.java:
```diff
@@ -78,25 +92,66 @@ public static int getAuthSessionsLimit(Config.Scope config) {
+        return config.getInt("stalledTimeoutInSeconds", defaultTimeout);
+    }
+
+    private void loadAuthSessionsFromRemoteCaches(final KeycloakSessionFactory sessionFactory, String cacheName, final int sessionsPerSegment, final int maxErrors) {
```
@ahus1 @mhajas can you get your feedback on this? We deprecated the remote store and the old logic didn't fetch the data from the remote store. Do you think it makes sense to add this now?

mposolda replied:
@pruivo @graziang I might miss some context and don't have latest knowledge of the infinispan/cluster/cross-dc layer in general, but AFAIK we never supported backup of `authenticationSessions` cache by the remote-store in the cross-dc environment in the old store (Infinispan distributed caches). We supported it for caches like `sessions`, `clientSessions`, `offlineSessions`, `loginFailures` etc, but not for `authenticationSessions` . So my vote is to not add it as part of this PR.

If I understand correctly, the related issues, which this PR solves ( #23881 and #32658 ) does not need `authenticationSessions` cache backed by the remote-store. So not needed to include in this PR IMO.

pruivo replied:
Also, I suspect it is broken since if it imports from an older version, the remote store will have `RootAuthenticationSessionEntity` stored. It will throw a `ClassCastException` because we store `SessionEntityWrapper` in the embedded cache.

mhajas replied:
> AFAIK we never supported backup of authenticationSessions cache by the remote-store in the cross-dc environment in the old store (Infinispan distributed caches).

Hmm this is interesting. It really seems initially, in tech preview of crossdc (kc 18), we didn't support/configure remote store for authentication sessions: https://github.com/keycloak/keycloak/blob/18.0.2/testsuite/integration-arquillian/servers/auth-server/jboss/common/jboss-cli/cross-dc-setup.cli but then with the first version of multi-site in Keycloak 24 we added it because we assumed it is necessary :/ https://github.com/keycloak/keycloak-benchmark/blob/6eb3efdb2007c5fa2b03223379e6d1bca20cdec9/provision/minikube/keycloak/config/kcb-infinispan-cache-remote-store-config.xml#L70-L96

Since we added support for it, I am afraid we will need to keep it part of this PR since we cannot do any breaking changes in minor release. 


> Also, I suspect it is broken since if it imports from an older version, the remote store will have RootAuthenticationSessionEntity stored. It will throw a ClassCastException because we store SessionEntityWrapper in the embedded cache.

@pruivo What do you mean by this? We are not changing the protostream representation here, so it should not be broken.

mhajas replied:
I am wondering, if it would be simpler to create a new AuthenticationSession provider without remote store support with fixes https://github.com/keycloak/keycloak/issues/23881 and https://github.com/keycloak/keycloak/issues/32658 and hide it behind a disabled_by_default feature and switch it to default in next major. WDYT?

pruivo replied:
You still need to do some migration at some point to convert from one class to another.

mhajas replied:
Yes, but we can do it in the next major version while keeping the previous behavior for minor releases

pruivo replied:
@mhajas so, do you want to delay this PR until Keycloak 27?

Keycloak 26.0.x has the remote store deprecated and stores `RootAuthenticationSessionEntity`. 
For Kecyloak 26.1.x (assuming the remote store is kept deprecated), if you want to merge this PR, you need to fetch the data from the remote store and replace/wrap it with the `SessionEntityWrapper`.

Am I wrong?

mhajas replied:
> @mhajas so, do you want to delay this PR until Keycloak 27?

No. What I am saying is that we now support storing authentication sessions in the remote store, therefore, we cannot remove support for it in this PR. This brings a lot of work needed to make this work with the ChangeLogBasedTransaction used for sessions as we are doing a lot of ugly things, like skipping of remote store etc. If I understand correctly for this reason we are also preloading the authentication sessions from remote store for this reason.

My suggestion is to introduce a new provider that would not have support for remote store at all that would not have issues https://github.com/keycloak/keycloak/issues/23881 and https://github.com/keycloak/keycloak/issues/32658. This new provider would be disabled by default and could be enabled by a feature like AUTHENTICATION_SESSIONS:v2 because it is breaking change as it does not support remote store. 

In Keycloak 27, we would switch it to be enabled by default and the old provider would be removed with the CACHE_EMBEDDED_REMOTE_STORE feature removed.


Does it make more sense now?

pruivo replied:
Users using the deprecated code will not have the issue fixed with that approach. We would have to duplicate the tests to enable the feature and ensure the tests are passing with the new implementation.

The number of issues I answered because users trying to do a rolling upgrade between KC25 -> KC26, and ProtoStream was complaining led me to believe they will never enable it :laughing:

mhajas replied:
> Users using the deprecated code will not have the issue fixed with that approach.

This issue is there probably since the beginning of Keycloak so I don't think it is that urgent to fix it asap.

> We would have to duplicate the tests to enable the feature and ensure the tests are passing with the new implementation.

That is true. However, we will need to find some good solution for this as we plan to utilize features more in future.

> The number of issues I answered because users trying to do a rolling upgrade between KC25 -> KC26, and ProtoStream was complaining led me to believe they will never enable it 😆

Then they will need to wait for fixing of the issues for the next major release. If they want the fix, they have the possibility to enable the feature.



We have 3 scenarios where authentication sessions are used. 
1. Single site using only embedded ISPN
2. Multi-site using the new Remote* implementation
3. Deprecated CACHE_EMBEDDED_REMOTE_STORE

This PR is trying to fix the bug in all 3. My assumption is that if we don't consider fixing it in 3 we could make it simpler and we could avoid some unexpected behavior, for example, when `InfinispanChangelogBasedTransaction` uses some unexpected `SKIP_CACHE_LOAD` flag to some operation that `AuthenticationSessionProvider` does not expect and similar.

But I must admit I didn't go through the PR thoroughly so maybe I am missing some other aspect of this. I am not saying my suggestion is the best option. What I am saying is that if we want to save some work to make remote-store work, there is an option. @pruivo If you want we can jump on a call and discuss it there.

pruivo replied:
I don't want to fix it for `CACHE_EMBEDDED_REMOTE_STORE` either, so importing it from the remote store is not worth it; we can just clear the remote store using a migration method. 

`InfinispanChangelogBasedTransaction` only decorates the cache for writes, where it writes to the embedded caches first and then to the external cluster. I'm optimistic here so I don't expect any issues. (although this is not efficient at all)

You know my desire of rewriting the embedded providers from scratch :smile:

mhajas replied:
Ok, I thought we are using `SKIP_CACHE_LOAD` also in the transaction while it is used only in the `UserSessionProvider`. Ok then. We can proceed with what we have. Cleaning remote store on migration sounds correct.

pruivo commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/InfinispanChangelogBasedTransaction.java:
```diff
@@ -237,6 +237,15 @@ private void replace(K key, MergedUpdate<V> task, SessionEntityWrapper<V> oldVer
+
+                if (session.shouldEvaluateRemoval() && task.shouldRemove(session)) {
+                    logger.debugf("Entity %s removed after evaluation", key);
+                    CacheDecorators.skipCacheStoreIfRemoteCacheIsEnabled(cache)
```
`writeCache` is decorated. 
```suggestion
                    writeCache
```

pruivo commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/remote/updater/authsession/RootAuthenticationSessionUpdater.java:
```diff
@@ -0,0 +1,241 @@
+    @Override
+    public Expiration computeExpiration() {
+        return new Expiration(
+                SessionTimeouts.getAuthSessionLifespanMS(realm, null, getValue()),
+                SessionTimeouts.getAuthSessionMaxIdleMS(realm, null, getValue()));
```
Arguments in the wrong order. 
```suggestion
                SessionTimeouts.getAuthSessionMaxIdleMS(realm, null, getValue()),
                SessionTimeouts.getAuthSessionLifespanMS(realm, null, getValue()));
```

pruivo commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/remote/RemoteInfinispanAuthenticationSessionProvider.java:
```diff
@@ -115,29 +113,8 @@ public void updateNonlocalSessionAuthNotes(AuthenticationSessionCompoundId compo
-
-    private RootAuthenticationSessionAdapter wrap(RealmModel realm, RootAuthenticationSessionEntity entity) {
-        return entity == null ? null : new RootAuthenticationSessionAdapter(session, new RootAuthenticationSessionUpdater(realm, entity, transaction), realm, authSessionsLimit);
+    public void remove(RootAuthenticationSessionEntity entity) {
```
unused method. remove?

mposolda reviewed (commented):
@graziang @pruivo Thanks for the PR and for the review!

Added one review comment/concern (answer to the previous comment by @pruivo). But I am rather re-assigning this PR to @ahus1 as it touches mostly the infinispan layer, so would prefer if it is reviewed/approved/merged by someone from SRE team.

pruivo commented:
Just a warning, this cannot be backported to the 26.0 branch. The embedded cache data changed from `RootAuthenticationSessionEntity` to `SessionEntityWrapper`.

mhajas commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanAuthenticationSessionProviderFactory.java:
```diff
@@ -78,25 +92,66 @@ public static int getAuthSessionsLimit(Config.Scope config) {
     }
 
+    private void loadAuthSessionsFromRemoteCaches(KeycloakSession session) {
+        for (String cacheName : remoteCacheInvoker.getRemoteCacheNames()) {
```
@graziang For this to work, we need to call `addRemoteCache` on `remoteCacheInvoker`, otherwise there are no caches in it.

graziang committed:
Auth session with changelog transaction

Closes #23881
Closes #32658

Signed-off-by: Giuseppe Graziano <g.graziano94@gmail.com>

Auth session with changelog transaction

Closes #23881
Closes #32658

Signed-off-by: Giuseppe Graziano <g.graziano94@gmail.com>

Auth session with changelog transaction

Closes #23881
Closes #32658

Signed-off-by: Giuseppe Graziano <g.graziano94@gmail.com>

graziang commented:
@pruivo @mhajas @mposolda Thank you for the review! I’ve applied your suggestions. My knowledge of Infinispan and related topics is limited, so please feel free to add any additional comments, suggestions, or commits as needed.

ahus1 reviewed (changes requested):
Thank you for this PR, this is going into the right direction. Some code cleanup needs to be done, and also some extra cache removals need to be eliminated. 

Overall the Authentication Sessions should also be present in the remote Infinispan to allow people to have a seamless handover between the datacenters. As an optimization we could document how to disable it, still I would want to add it. 

As now a session wrapper is used as Pedro pointed out, this is incompatible with previous versions. So in a migration all existing cache entries for authentication sessions need to be either removed or converted. 

The details:

Class AuthenticationSessionTransaction is not used and can be IMHO removed.

InfinispanAuthenticationSessionProviderFactory#loadAuthSessionsFromRemoteCaches is a private method that is not being called. Given the discussion above this also hasn't been done before. With the read-through for any missing entries, loading all entries shouldn't be necessary. 

There is a mechanism in InfinispanKeycloakTransaction that prevented removals from Infinispan when the entry was added during the current transaction. With this recent change, this is no longer working and it would trigger extra removals possibly to remove Keycloak instances in the local cluster due to that. This would increase the latency of logouts and client credential grants: 

* When logging out for example from the admin console, InfinispanChangelogBasedTransaction#runOperationInCluster is called twice to remove an authentication session when using the embedded cache. And once when using the remove cache it is called in RemoteChangeLogTransaction#doCommit with DELETE. I assume this session is created within the current request. As it hasn't been sent to Infinispan, no remove should be called, which will make this faster. 

* When running a client credential grant, a remove to Infinispan is triggered (when using embedded cache and also when remote Infinispan)

graziang committed:
Follow-up Auth session with changelog transaction

Closes #23881
Closes #32658

Signed-off-by: Giuseppe Graziano <g.graziano94@gmail.com>

graziang commented:
@pruivo @mhajas @ahus1 Thanks for all the comments and reviews. I’ve added a commit with the following changes:

- Removed the `AuthenticationSessionTransaction` class.
- Removed the `InfinispanAuthenticationSessionProviderFactory#loadAuthSessionsFromRemoteCaches` method.
- Added remote store cache cleanup in the 26.1 migration.
- Added a mechanism to prevent storing the entity in cache if it’s created and deleted within the same transaction:
  - For the embedded cache in `InfinispanChangelogBasedTransaction`, skip the cache tasks if the first task is an ADD and the last is a REMOVE.
  - For remote cache, added a mechanism in the `BaseUpdater` to mark the entity as transient if it’s created and deleted in the same transaction.

pruivo commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanAuthenticationSessionProvider.java:
```diff
@@ -162,36 +170,27 @@ public RootAuthenticationSessionModel getRootAuthenticationSession(RealmModel re
-        }
+    @Override
+    public void migrate(String modelVersion) {
+        InfinispanConnectionProvider infinispanConnectionProvider = session.getProvider(InfinispanConnectionProvider.class);
```
check the `modelVersion` the cache is clear in every version change. 
```java
if ("26.1.0".equals(modelVersion))
```

graziang replied:
Done!

pruivo commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanAuthenticationSessionProvider.java:
```diff
@@ -162,36 +170,27 @@ public RootAuthenticationSessionModel getRootAuthenticationSession(RealmModel re
+    public void migrate(String modelVersion) {
+        InfinispanConnectionProvider infinispanConnectionProvider = session.getProvider(InfinispanConnectionProvider.class);
+        Cache<String, RootAuthenticationSessionEntity> authSessionsCache = infinispanConnectionProvider.getCache(InfinispanConnectionProvider.AUTHENTICATION_SESSIONS_CACHE_NAME);
+        ComponentRegistry.componentOf(authSessionsCache, PersistenceManager.class).clearAllStores(PersistenceManager.AccessMode.BOTH);
```
Wait until clear is completed
```suggestion
        CompletionStages.join(ComponentRegistry.componentOf(authSessionsCache, PersistenceManager.class).clearAllStores(PersistenceManager.AccessMode.BOTH));
```

graziang replied:
Done!

pruivo commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanAuthenticationSessionProviderFactory.java:
```diff
@@ -67,6 +74,11 @@ public class InfinispanAuthenticationSessionProviderFactory implements Authentic
 
+    private RemoteCacheInvoker remoteCacheInvoker;
+    SerializeExecutionsByKey<String> serializer = new SerializeExecutionsByKey<>();
+
+
+
```
nitpicking, too many new lines :smile:

graziang replied:
Done!

pruivo commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/remote/updater/BaseUpdater.java:
```diff
@@ -75,9 +76,17 @@ public final boolean isReadOnly() {
+        if (isCreated()) {
+            transientEntity = true;
+        }
         state = UpdaterState.DELETED;
```
this is in an extreme level of nitpicking, but the boolean needs 8 more bytes in memory. I would suggest to add the value `UpdaterState.TRANSIENT` and change this to
```suggestion
        state = isCreated() ? UpdaterState.TRANSIENT : UpdaterState.DELETED;
```
JOL output with COOPS (compressed references) & CCPS (compressed classes)

```
org.keycloak.models.sessions.infinispan.changes.remote.updater.BaseUpdater object internals:
OFF  SZ           TYPE DESCRIPTION                   VALUE
  0   8                (object header: mark)         N/A
  8   4                (object header: class)        N/A
 12   1        boolean BaseUpdater.transientEntity   N/A
 13   3                (alignment/padding gap)       
 16   8           long BaseUpdater.versionRead       N/A
 24   4              K BaseUpdater.cacheKey          N/A
 28   4              V BaseUpdater.cacheValue        N/A
 32   4   UpdaterState BaseUpdater.state             N/A
 36   4                (object alignment gap)        
Instance size: 40 bytes
Space losses: 3 bytes internal + 4 bytes external = 7 bytes total
```
```
org.keycloak.models.sessions.infinispan.changes.remote.updater.BaseUpdater object internals:
OFF  SZ           TYPE DESCRIPTION               VALUE
  0   8                (object header: mark)     N/A
  8   4                (object header: class)    N/A
 12   4              K BaseUpdater.cacheKey      N/A
 16   8           long BaseUpdater.versionRead   N/A
 24   4              V BaseUpdater.cacheValue    N/A
 28   4   UpdaterState BaseUpdater.state         N/A
Instance size: 32 bytes
```

graziang replied:
Done!

pruivo reviewed (changes requested):
LGTM, the changes in `migrate()` method are required and the remaining comments are optional. Nice work @graziang

graziang committed:
Follow-up Auth session with changelog transaction

Closes #23881
Closes #32658

Signed-off-by: Giuseppe Graziano <g.graziano94@gmail.com>

pruivo reviewed (approved):
LGTM! Thanks @graziang !

ahus1 reviewed (approved):
Thank you for this extensive change around authentication sessions! Great to see progress!
